src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/JsonCollectionConverter.cs

```diff
@@ -143,8 +143,6 @@ protected static JsonConverter<TElement> GetElementConverter(ref WriteStack stat
                     {
                         ThrowHelper.ThrowJsonException_DeserializeUnableToConvertValue(Type);
                     }
-
-                    state.Current.JsonPropertyInfo = elementTypeInfo.PropertyInfoForTypeInfo;
```

Comment:
The crux of the fix is that this initialization was happening too early -- if the converter was run as part of a polymorphic re-entrancy this initialization would never be hit. I checked the other polymorphic converters and none have the same issue.